src/utp_wrapper.rs

```diff
@@ -32,15 +37,40 @@ impl UtpWrapper {
                 .spawn(move || {
                     let mut socket = socket;
                     socket.set_read_timeout(Some(CHECK_FOR_NEW_WRITES_INTERVAL_MS));
+                    let heartbeat_msg = unwrap_result!(serialise(&CrustMsg::Heartbeat));
+                    let mut heartbeat_deadline = Instant::now() + heartbeat_timeout;
+                    let mut inactivity_deadline = Instant::now() + inactivity_timeout;
+                    let update_heartbeat = |heartbeat_deadline: &mut Instant, inactivity_deadline: &mut Instant| {
+                        let now = Instant::now();
+                        *heartbeat_deadline = now + heartbeat_timeout;
+                        *inactivity_deadline = now + inactivity_timeout;
+                    };
                     'read_write: loop {
                         let mut buf = [0; BUFFER_SIZE];
+                        {
+                            let now = Instant::now();
+                            if now > inactivity_deadline {
+                                info!("Stale connection. Dropping...");
+                                break;
+                            }
+                            if now > heartbeat_deadline {
+                                if let Err(e) = socket.send_to(&heartbeat_msg) {
+                                    error!("Error sending: {:?}", e);
+                                    break;
+                                }
+                                update_heartbeat(&mut heartbeat_deadline,
+                                                 &mut inactivity_deadline);
```

Comment:
The `inactivity_timeout` is supposed to make us ignore the underlying protocol rules and drop the connection if we don't ever detect activity within the time given.

I just end up modelling this behaviour, but as you noticed, this timeout will only be reached when thread goes to sleep for a long time. Maybe I should just remove this variable. What do you think?

I also update the `inactivity_timeout` when a message is sent because this will also trigger the uTP socket to exchange messages with the remote peer.